File: gp/src/main/java/uk/nhs/wales/gppractice/model/FhirOperationOutcome.java
package uk.nhs.wales.gppractice.model;

import org.hl7.fhir.r4.model.OperationOutcome;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class FhirOperationOutcome {
    private final String resourceType = "OperationOutcome";
    private List<Map<String, Object>> issue;

    public String getResourceType() { 
        return resourceType; 
    }
    
    public List<Map<String, Object>> getIssue() { 
        return issue; 
    }
    
    public void setIssue(List<Map<String, Object>> issue) { 
        this.issue = issue; 
    }

    public static FhirOperationOutcome success(String message) {
        return create("information", "informational", message);
    }

    public static FhirOperationOutcome error(String message) {
        return create("error", "invalid", message);
    }

    public static FhirOperationOutcome fromOperationOutcome(OperationOutcome operationOutcome) {
        FhirOperationOutcome outcome = new FhirOperationOutcome();
        
        if (operationOutcome.hasIssue()) {
            outcome.setIssue(
                operationOutcome.getIssue().stream()
                    .map(issue -> Map.of(
                        "severity", issue.hasSeverity() ? issue.getSeverity().toCode() : "error",
                        "code", issue.hasCode() ? issue.getCode().toCode() : "unknown",
                        "details", Map.of(
                            "text", issue.hasDiagnostics() ? issue.getDiagnostics() : 
                                   (issue.hasDetails() && issue.getDetails().hasText() ? 
                                    issue.getDetails().getText() : "Validation failed")
                        )
                    ))
                    .collect(Collectors.toList())
            );
        } else {
            outcome.setIssue(List.of(Map.of(
                "severity", "error",
                "code", "unknown",
                "details", Map.of("text", "Validation failed")
            )));
        }
        
        return outcome;
    }

    private static FhirOperationOutcome create(String severity, String code, String message) {
        FhirOperationOutcome outcome = new FhirOperationOutcome();
        outcome.setIssue(List.of(Map.of(
                "severity", severity,
                "code", code,
                "details", Map.of("text", message)
        )));
        return outcome;
    }
}

File: src/Models/FhirOperationOutcome.cs
using System.Text.Json.Serialization;
using Hl7.Fhir.Model;

namespace Uk.HealthTechWales.GpPractice.Models;

public class FhirOperationOutcome
{
    [JsonPropertyName("resourceType")]
    public string ResourceType { get; } = "OperationOutcome";

    [JsonPropertyName("issue")]
    public List<Dictionary<string, object>>? Issue { get; set; }

    public static FhirOperationOutcome Success(string message)
    {
        return Create("information", "informational", message);
    }

    public static FhirOperationOutcome Error(string message)
    {
        return Create("error", "invalid", message);
    }

    public static FhirOperationOutcome FromOperationOutcome(OperationOutcome operationOutcome)
    {
        var outcome = new FhirOperationOutcome();

        if (operationOutcome.Issue != null && operationOutcome.Issue.Any())
        {
            outcome.Issue = operationOutcome.Issue
                .Select(issue => new Dictionary<string, object>
                {
                    ["severity"] = issue.Severity?.ToString()?.ToLower() ?? "error",
                    ["code"] = issue.Code?.ToString()?.ToLower() ?? "unknown",
                    ["details"] = new Dictionary<string, object>
                    {
                        ["text"] = issue.Diagnostics ??
                                  (issue.Details?.Text ?? "Validation failed")
                    }
                })
                .ToList();
        }
        else
        {
            outcome.Issue = new List<Dictionary<string, object>>
            {
                new()
                {
                    ["severity"] = "error",
                    ["code"] = "unknown",
                    ["details"] = new Dictionary<string, object> { ["text"] = "Validation failed" }
                }
            };
        }

        return outcome;
    }

    private static FhirOperationOutcome Create(string severity, string code, string message)
    {
        return new FhirOperationOutcome
        {
            Issue = new List<Dictionary<string, object>>
            {
                new()
                {
                    ["severity"] = severity,
                    ["code"] = code,
                    ["details"] = new Dictionary<string, object> { ["text"] = message }
                }
            }
        };
    }
}
